daly-bms-uart.cpp

```diff
@@ -39,63 +46,417 @@ bool Daly_BMS_UART::Init()
+    {
+        return false;
+    }
+    this->sendCommand(COMMAND::CELL_VOLTAGES);
 
-    if (!receiveBytes())
+    for (size_t i = 0; i <= ceil(get.numberOfCells / 3); i++)
     {
-#ifdef DALY_BMS_DEBUG
-        Serial.printf("<DALY-BMS DEBUG> Receive failed, min/max cell values won't be modified!\n");
+        if (!this->receiveBytes())
+        {
+#ifdef DEBUG_SERIAL
+            DEBUG_SERIAL.print("<DALY-BMS DEBUG> Receive failed, Cell Voltages won't be modified!\n");
+#endif
+            return false;
+        }
+
+        for (size_t i = 0; i < 3; i++)
+        {
+
+#ifdef DEBUG_SERIAL
+            DEBUG_SERIAL.print("<DALY-BMS DEBUG> Frame No.: " + (String)this->my_rxBuffer[4]);
+            DEBUG_SERIAL.println(" Cell No: " + (String)(cellNo + 1) + ". " + (String)((this->my_rxBuffer[5 + i + i] << 8) | this->my_rxBuffer[6 + i + i]) + "mV");
 #endif
+
+            get.cellVmV[cellNo] = (this->my_rxBuffer[5 + i + i] << 8) | this->my_rxBuffer[6 + i + i];
+            cellNo++;
+            if (cellNo >= get.numberOfCells)
+                break;
+        }
+    }
+    return true;
+}
+
+bool Daly_BMS_UART::getCellTemperature() // 0x96
+{
+    int sensorNo = 0;
+
+    if (get.numOfTempSensors < 1 && get.numOfTempSensors >= 16)
+    {
         return false;
     }
+    this->sendCommand(COMMAND::CELL_TEMPERATURE);
+
+    for (size_t i = 0; i <= ceil(get.numOfTempSensors / 7); i++)
+    {
+
+        if (!this->receiveBytes())
+        {
+#ifdef DEBUG_SERIAL
+            DEBUG_SERIAL.print("<DALY-BMS DEBUG> Receive failed, Cell Temperatures won't be modified!\n");
+#endif
+            return false;
+        }
 
-    maxCellV = (float)((this->my_rxBuffer[4] << 8) | this->my_rxBuffer[5]) / 1000; // Given in mV, convert to V
-    maxCellVNum = this->my_rxBuffer[6];
-    minCellV = (float)((this->my_rxBuffer[7] << 8) | this->my_rxBuffer[8]) / 1000; // Given in mV, convert to V
-    minCellVNum = this->my_rxBuffer[9];
+        for (size_t i = 0; i < 7; i++)
+        {
+
+#ifdef DEBUG_SERIAL
+            DEBUG_SERIAL.print("<DALY-BMS DEBUG> Frame No.: " + (String)this->my_rxBuffer[4]);
+            DEBUG_SERIAL.println(" Sensor No: " + (String)(sensorNo + 1) + ". " + String(this->my_rxBuffer[5 + i] - 40) + "°C");
+#endif
+
+            get.cellTemperature[sensorNo] = (this->my_rxBuffer[5 + i] - 40);
+            sensorNo++;
+            if (sensorNo + 1 >= get.numOfTempSensors)
+                break;
+        }
+    }
+    return true;
+}
+
+bool Daly_BMS_UART::getCellBalanceState() // 0x97
+{
+    int cellBalance = 0;
+    int cellBit = 0;
+
+    if (get.numberOfCells < 1 && get.numberOfCells >= 48)
+    {
+        return false;
+    }
+    this->sendCommand(COMMAND::CELL_BALANCE_STATE);
+
+    if (!this->receiveBytes())
+    {
+#ifdef DEBUG_SERIAL
+        DEBUG_SERIAL.println("<DALY-BMS DEBUG> Receive failed, Cell Balance State won't be modified!\n");
+#endif
+        return false;
+    }
+
+    for (size_t i = 0; i < 6; i++)
```

Comment:
@softwarecrash  Can you add a comment here explaining what's going on with these for loops, it's not immediately obvious

Reply:
@maland16 sure,
the bms send for every cell a balance active bit, the inner for loop iterate a byte to bitwise and set the corespond bool flag on the array, the outer for iterate over the bytes in the bms message so it runs 6*8=48 the maximum numbers of cells. hope you can now comprehend it?